Add Jenkins docs (#1416)
theforeman/foreman-infra: docs/jenkins.md

Each plugin should have a job defined in JJB that calls test_plugin_matrix here: https://ci.theforeman.org/view/Plugins/

#### Foreman plugin PR testing

To test pull requests, a separate job is used that also takes the PR details: https://ci.theforeman.org/view/Plugins/job/test_plugin_pull_request/

#### Adding a new Foreman plugin

For a plugin "foreman_example", first create a job that tests the main (master or develop) branch.

* ensure plugin tests (if any) run when `rake jenkins:unit` is called, see [the example plugin](https://github.com/theforeman/foreman_plugin_template/) and [testing a plugin](https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Testing) for help
* create a foreman_example.yaml file in [foreman-infra/JJB](https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files/theforeman.org/yaml/jobs/plugins)
  * This will create a "test_plugin_foreman_example_master" job in Jenkins to test the master branch.
* ensure the job is green by fixing bugs, installing dependencies etc.

> * add hook to GitHub repo, see [GitHub repo hook](https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop)

Reviewer: This link still refers to the wiki, but this document is supposed to replace it.

Perhaps this should also at the org admin part since adding the Bots team is also a Github repo action.

Each plugin should have a job defined in JJB that calls test_proxy_plugin_matrix here: https://ci.theforeman.org/view/Plugins/

#### Smart proxy plugin PR testing

To test pull requests, a separate job is used that also takes the PR details: https://ci.theforeman.org/view/Plugins/job/test_proxy_plugin_pull_request/

### Adding a new smart proxy plugin

For a plugin "smart_proxy_example", first create a job that tests the main (master or develop) branch.

* ensure plugin tests run when doing `bundle install` and `rake test`, see [testing a plugin](https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Smart-Proxy_Plugin#Testing) for help
* if different branches rely on different versions of smart proxy, specify `:branch` in Gemfile on those branches
* create a smart_proxy_example.yaml file in foreman-infra/JJB
* https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files/theforeman.org/yaml/jobs/plugins
* This will create a "test_proxy_plugin_smart_proxy_example_master" job in Jenkins to test the master branch.
* ensure the job is green by fixing bugs, installing dependencies etc.

> * add hook to GitHub repo, see [GitHub repo hook](https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop)

Reviewer: Same here

Next, set up PR testing for the plugin:

* add template to foreman-infra/puppet/modules/slave/templates, update branch, job names etc.

An org admin must then:

* add the repo to the [Bots team](https://github.com/orgs/theforeman/teams/bots/repositories) with **write** access

## Job configurations

### Testing develop

All repos with an associated job that tests their master/develop branch should have a hook added to the repo to trigger immediate builds.

To set up the hook, an org/repo admin must:

* View the repository settings
* Click _Webhooks

>   * _Add service_ dropdown, select _Jenkins (Git plugin)_

Reviewer: This no longer works since GH doesn't have services anymore. It's replaced with:
* Payload URL: https://ci.theforeman.org/github-webhook/
* Content Type: application/json (default)
* Secret: the secret (which I think we haven't stored in a shared secret)
* Just the push event

## Jenkins Nodes

### Configuration management

All nodes are maintained through our own Foreman instance using Puppet.  The Foreman instance has a host group called "Builders" and "Builders/Debian" which have the "slave" and other classes assigned to them. Debian machines have additional permissions to push Debian packages.

https://github.com/theforeman/foreman-infra/tree/master/puppet/modules contains the source for all Puppet modules.

### Node requirements

* CentOS 7
  * Clean, minimal base installation or the option to reinstall it
* 2GB of RAM per vCPU (4 vCPU + 8GB RAM is typical)
* 60GB disk (minimum), SSD preferred
* ~20GB/month bandwidth
* Public facing IP address
* Root access

### Configuring a new node

For Enterprise Linux:

* Ensure EPEL is configured: [epel-release](https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)
* Ensure yum.pl.com is configured: [puppet6-release](https://yum.puppet.com/puppet6/puppet6-release-el-7.noarch.rpm)
* `yum -y install puppet-agent`
* `echo "server = puppetmaster.theforeman.org" >> /etc/puppetlabs/puppet/puppet.conf`

> * ensure hostname is set node0X.jenkins.<provider>.theforeman.org where <provider> is osuosl or aws for example

Reviewer: And that the actual DNS record exists

* Make the `puppet` command available: `source /etc/profile.d/puppet-agent.sh`
* `puppet ssl bootstrap`
* Sign the certificate on the puppetmaster or via Foreman
* `puppet agent --test`
* Set the host group to "Builders" in Foreman
* Run `puppet agent --test` twice (second run is important, due to the rvm module behaviour)

For Debian:

* Ensure apt.pl.com is configured: [puppet6-release](https://apt.puppetlabs.com/puppet6-release-buster.deb)
* `apt update && apt install puppet-agent`
* `echo "server = puppetmaster.theforeman.org" >> /etc/puppetlabs/puppet/puppet.conf`
* Make the `puppet` command available: `source /etc/profile.d/puppet-agent.sh`

> * `export PATH=/opt/puppetlabs/bin:$PATH`

Reviewer: This is redundant if you've sourced `/etc/profile.d/puppet-agent.sh`

#### Foreman plugin PR testing

To test pull requests, a separate job is used that also takes the PR details: https://ci.theforeman.org/view/Plugins/job/test_plugin_pull_request/

#### Adding a new Foreman plugin

For a plugin "foreman_example", first create a job that tests the main (master or develop) branch.

* ensure plugin tests (if any) run when `rake jenkins:unit` is called, see [the example plugin](https://github.com/theforeman/foreman_plugin_template/) and [testing a plugin](https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Testing) for help
* create a foreman_example.yaml file in [foreman-infra/JJB](https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files/theforeman.org/yaml/jobs/plugins)
  * This will create a "test_plugin_foreman_example_master" job in Jenkins to test the master branch.
* ensure the job is green by fixing bugs, installing dependencies etc.
* add hook to GitHub repo, see [GitHub repo hook](https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop)

An org admin must then:

> * add the repo to the [Bots team](https://github.com/orgs/theforeman/teams/bots/repositories) with +write+ access

Reviewer: ```suggestion
* add the repo to the [Bots team](https://github.com/orgs/theforeman/teams/bots/repositories) with **write** access
```

* create a smart_proxy_example.yaml file in foreman-infra/JJB
* https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files/theforeman.org/yaml/jobs/plugins
* This will create a "test_proxy_plugin_smart_proxy_example_master" job in Jenkins to test the master branch.
* ensure the job is green by fixing bugs, installing dependencies etc.
* add hook to GitHub repo, see [GitHub repo hook](https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop)

Next, set up PR testing for the plugin:

* add template to foreman-infra/puppet/modules/slave/templates, update branch, job names etc.

An org admin must then:

* add the repo to the [Bots team](https://github.com/orgs/theforeman/teams/bots/repositories) with **write** access

## Job configurations

### Testing develop

All repos with an associated job that tests their master/develop branch should have a hook added to the repo to trigger immediate builds.

To set up the hook, an org/repo admin must:

* View the repository settings
* Click _Webhooks & Services_
  * _Add service_ dropdown, select _Jenkins (Git plugin)_

> * For the URL, enter `https://ci.theforeman.org/`

Reviewer: I think this is outdated since Github no longer has Services. There's only webhooks now.

PR jobs should be set up identically to primary branch tests, except for the SCM (which checks out `${sha1}`) and to add the GHPRB trigger (see the `github_pr` macro in JJB).

To set up the hook, an org/repo admin goes to the repository settings, then Webhooks & Services and adds a webhook with these settings:

* Payload URL: `https://ci.theforeman.org/ghprbhook/`
* Content type: `application/json`
* Secret: _redacted_
* Events: _Let me select individual events_, _Pull request_, _Issue comment_

An org admin must then change the org teams:

* Add the repository to the [Bots team](https://github.com/orgs/theforeman/teams/bots/repositories) with **write** access

### CentOS QA service

CentOS offer access to their [QA CI infrastructure](https://wiki.centos.org/QaWiki/CI) to Foreman and other projects, which gives on-demand access to run tests on physical hardware running CentOS.  It consists of three main components:

* [ci.centos.org](https://ci.centos.org) - a Jenkins instance which we can manage jobs on
* foreman@slave01 - a user account on a slave which our jobs all run on

> * [Duffy](https://wiki.centos.org/QaWiki/CI/Duffy) - on-demand provisioning of physical test servers running CentOS, available from foreman`slave01 jobs

Reviewer: The slave01 part it no really correct anymore. I think we now limit this the required secret to OSUOSL nodes.

Reviewer 2: That's the account on ci.centos.org infra. It at least that's how I read that section and it matches my knowledge of the infra.

Reviewer 2: https://ci.centos.org/computer/foreman-ci-slave01/

Reviewer: ah right

## Managing jobs

> Jenkins itself is deployed onto one master VM from foreman-infra.  Jobs are maintained in two ways:
> 
> * Updated by hand - some developers have accounts to log in and modify jobs
> * Jenkins Job Builder - a set of YAML configuration files in foreman-infra that generate jobs
> 
> New jobs should be written via Jenkins Job Builder if possible.

Reviewer: ```suggestion
Jenkins itself is deployed onto one master VM from foreman-infra.  Jobs are maintained via Jenkins Job Builder - a set of YAML configuration files in foreman-infra that generate jobs. Jobs are synced via Puppet and unmanaged jobs are purged daily.
```

### Jenkins Job Builder

[Jenkins Job Builder](https://docs.openstack.org/infra/jenkins-job-builder/) (JJB) is an OpenStack tool for generating Jenkins job definitions (an XML file) from a set of YAML job descriptions, which we store in version control.

* [JJB YAML files in foreman-infra](https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files)

Puppet deploys these onto our Jenkins server (a recursive file copy) and when they change, it runs the JJB tool to update the jobs in the live instance.  It also refreshes them daily to overwrite manual changes.

Useful resources:

* [Job definitions, templates etc.](https://docs.openstack.org/infra/jenkins-job-builder/definition.html)
* [Modules, e.g. SCM, publishers, builders](https://docs.openstack.org/infra/jenkins-job-builder/definition.html#modules)

## Jenkins Nodes

### Configuration management

> All nodes are maintained through our own Foreman instance using Puppet.  The Foreman instance has a host group called "Base/Builders" and "Base/Builders/Red Hat" and "Base/Builders/Debian" which have the "node" and other classes assigned to them.

Reviewer: ```suggestion
All nodes are maintained through our own Foreman instance using Puppet.  The Foreman instance has a host group called "Builders" and "Builders/Debian" which have the "slave" and other classes assigned to them. Debian machines have additional permissions to push Debian packages.
```

https://github.com/theforeman/foreman-infra/tree/master/puppet/modules contains the source for all Puppet modules.

### Node requirements

* CentOS 7
  * Clean, minimal base installation or the option to reinstall it
* 2GB of RAM per vCPU (4 vCPU + 8GB RAM is typical)
* 60GB disk (minimum), SSD preferred
* ~20GB/month bandwidth
* Public facing IP address
* Root access

### Configuring a new node

For Enterprise Linux:

* Ensure EPEL is configured: [epel-release](https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)
* Ensure yum.pl.com is configured: [puppet6-release](https://yum.puppet.com/puppet6/puppet6-release-el-7.noarch.rpm)
* `yum -y install puppet-agent`
* `echo "server = puppetmaster.theforeman.org" >> /etc/puppetlabs/puppet/puppet.conf`

> * Relog into the machine so that the 'puppet' command is available

Reviewer: You can also `source /etc/profile.d/puppet-agent.sh`.
```suggestion
* Make the `puppet` command available: `source /etc/profile.d/puppet-agent.sh`
```

> * `puppet agent --test`

Reviewer: If we're on Puppet 6, let's use the current best practice
```suggestion
* `puppet ssl bootstrap`
```

* Sign the certificate on the puppetmaster or via Foreman
* `puppet agent --test`

> * Set the host group to "Base/Builders/Red Hat" in Foreman

Reviewer: ```suggestion
* Set the host group to "Builders" in Foreman
```

* Run `puppet agent --test` twice (second run is important, due to the rvm module behaviour)

For Debian:

* Ensure apt.pl.com is configured: [puppet6-release](https://apt.puppetlabs.com/puppet6-release-buster.deb)

> * `apt get puppet-agent`

Reviewer: ```suggestion
* `apt update && apt install puppet-agent`
```

* `echo "server = puppetmaster.theforeman.org" >> /etc/puppetlabs/puppet/puppet.conf`

> * Relog into the machine so that the 'puppet' command is available
> * `export PATH=/opt/puppetlabs/bin:$PATH`

Reviewer: ```suggestion
* Make the `puppet` command available: `source /etc/profile.d/puppet-agent.sh`
```

> * ensure hostname is set node0X.jenkins.<provider>.theforeman.org where <provider> is osuosl or aws for example

Reviewer: This also goes for Enterprise Linux.

> * `puppet agent --test`

Reviewer: ```suggestion
* `puppet ssl bootstrap`
```

* Sign the certificate on the puppetmaster or via Foreman
* `puppet agent --test`

> * Set the host group to "Base/Builders/Debian" in Foreman

Reviewer: ```suggestion
* Set the host group to "Builders/Debian" in Foreman
```

# Jenkins

The Foreman project maintains its own Jenkins instance for continuous integration at https://ci.theforeman.org/

It runs the following types of tests:

* Unit tests on the main develop and all supported stable branches of core Foreman projects
* Pull request tests against core Foreman projects
* Unit and pull request tests for many Foreman plugins
* Nightly and release package building and release processes
* Package smoke and integration testing
* Package jobs for Pulpcore

## Quick reference for maintainers

Current PR test jobs (used on Foreman itself) support these commands:

* `ok to test` - run tests for an unknown user, if the code within the patch is not malicious
* `[test STATUS-NAME]`, e.g. `[test foreman]` to re-run a particular set of tests

## Quick reference for plugin maintainers

### Foreman plugin testing

> Foreman plugins are tested by adding the plugin to a Foreman checkout and running core tests, so it checks that existing behaviours still work and new plugin tests are run too.  The "test_plugin_matrix job":https://ci.theforeman.org/job/test_plugin_matrix/ copies the core jobs, but adds a plugin from a given git repo/branch and is usually used to test plugins in a generic way.

Reviewer 2: ```suggestion
Foreman plugins are tested by adding the plugin to a Foreman checkout and running core tests, so it checks that existing behaviours still work and new plugin tests are run too.  The [test_plugin_matrix job](https://ci.theforeman.org/job/test_plugin_matrix/) copies the core jobs, but adds a plugin from a given git repo/branch and is usually used to test plugins in a generic way.
```

Each plugin should have a job defined in JJB that calls test_plugin_matrix here: https://ci.theforeman.org/view/Plugins/

#### Foreman plugin PR testing

To test pull requests, a separate job is used that also takes the PR details: https://ci.theforeman.org/view/Plugins/job/test_plugin_pull_request/

#### Adding a new Foreman plugin

For a plugin "foreman_example", first create a job that tests the main (master or develop) branch.

> * ensure plugin tests (if any) run when `rake jenkins:unit` is called, see "the example plugin":https://github.com/theforeman/foreman_plugin_template/ and "testing a plugin":https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Testing for help

Reviewer 2: ```suggestion
* ensure plugin tests (if any) run when `rake jenkins:unit` is called, see [the example plugin](https://github.com/theforeman/foreman_plugin_template/) and [testing a plugin](https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Testing) for help
```

> * create a foreman_example.yaml file in foreman-infra/JJB

Reviewer 2: in where? :)

>   * https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files/theforeman.org/yaml/jobs/plugins

Reviewer 2: probably in HERE, so maybe the two lines should be folded together

  * This will create a "test_plugin_foreman_example_master" job in Jenkins to test the master branch.
* ensure the job is green by fixing bugs, installing dependencies etc.

> * add hook to GitHub repo, see "GitHub repo hook":https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop

Reviewer 2: ```suggestion
* add hook to GitHub repo, see [GitHub repo hook](https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop)
```

the link is wrong tho, it should be in this doc, right?

An org admin must then:

> * add the repo to the "Bots team":https://github.com/orgs/theforeman/teams/bots/repositories with +write+ access

Reviewer 2: ```suggestion
* add the repo to the [Bots team](https://github.com/orgs/theforeman/teams/bots/repositories) with `write` access
```

### Smart proxy plugin testing

> Proxy plugins are tested like ordinary gems with tests run entirely from the plugin directory, installing the smart proxy as a dependency (via bundler's git support).  The "test_proxy_plugin_matrix job":https://ci.theforeman.org/job/test_proxy_plugin_matrix/ is usually used to test plugins in a generic way.

Reviewer 2: ```suggestion
Proxy plugins are tested like ordinary gems with tests run entirely from the plugin directory, installing the smart proxy as a dependency (via bundler's git support).  The [test_proxy_plugin_matrix job](https://ci.theforeman.org/job/test_proxy_plugin_matrix/) is usually used to test plugins in a generic way.
```

Each plugin should have a job defined in JJB that calls test_proxy_plugin_matrix here: https://ci.theforeman.org/view/Plugins/

#### Smart proxy plugin PR testing

To test pull requests, a separate job is used that also takes the PR details: https://ci.theforeman.org/view/Plugins/job/test_proxy_plugin_pull_request/

> To set this up for a plugin, make the template + manifest addition as noted in the PR section below to configure the PR scanner.  Also enable the prprocessor webhook on the GitHub repo for immediate builds.

Reviewer 2: E_NO_PR_SCANNER

### Adding a new smart proxy plugin

For a plugin "smart_proxy_example", first create a job that tests the main (master or develop) branch.

> * ensure plugin tests run when doing `bundle install` and `rake test`, see "testing a plugin":https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Smart-Proxy_Plugin#Testing for help

Reviewer 2: ```suggestion
* ensure plugin tests run when doing `bundle install` and `rake test`, see [testing a plugin](https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Smart-Proxy_Plugin#Testing) for help
```

#* if different branches rely on different versions of smart proxy, specify `:branch` in Gemfile on those branches
* create a smart_proxy_example.yaml file in foreman-infra/JJB
#* https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files/theforeman.org/yaml/jobs/plugins

> #* This will create a "test_proxy_plugin_smart_proxy_example_master" job in Jenkins to test the master branch.

Reviewer 2: this formatting is weird, what's `#*`?

* ensure the job is green by fixing bugs, installing dependencies etc.

> * add hook to GitHub repo, see "GitHub repo hook":https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop

Reviewer 2: as above, this should be an in-doc link

Next, set up PR testing for the plugin:

> * add template to foreman-infra/puppet/modules/slave/templates, update branch, job names etc.

Reviewer 2: nope

Reviewer: Not sure why this was marked as resolved because this is no longer relevant and can be dropped.

> * add project to slave::pr_test_config list in foreman-infra/puppet/modules/slave/manifests/init.pp

Reviewer 2: nopenope

> * add project to [[PrProcessor]] "repos.yaml":https://github.com/theforeman/prprocessor/blob/master/config/repos.yaml config file with "pr_scanner: true"

Reviewer 2: I guess adding it to prprocessor app is a good idea, but otherwise? @ekohl?

Reviewer: yes

An org admin must then:

> * add the repo to the "Bots team":https://github.com/orgs/theforeman/teams/bots/repositories with +write+ access

Reviewer 2: ```suggestion
* add the repo to the [Bots team](https://github.com/orgs/theforeman/teams/bots/repositories) with `write` access
```

## Job configurations

### Testing develop

All repos with an associated job that tests their master/develop branch should have a hook added to the repo to trigger immediate builds.

To set up the hook, an org/repo admin must:

* View the repository settings
* Click _Webhooks & Services_
  * _Add service_ dropdown, select _Jenkins (Git plugin)_
* For the URL, enter `https://ci.theforeman.org/`

### Pull request testing

Core Foreman projects have GitHub pull requests tested on our Jenkins instance so it's identical to the way we test the primary development branches themselves.  Simpler and quieter projects (such as installer modules) should use Travis CI which supports PR testing and reduces the load on our own infrastructure.

Every project that needs PR testing has at least two Jenkins jobs.  Taking core Foreman as an example:

> * Test job for the main development branch (develop or master): "test_develop":https://ci.theforeman.org/job/test_develop/

Reviewer 2: ```suggestion
* Test job for the main development branch (develop or master): [test_develop](https://ci.theforeman.org/job/test_develop/)
```

> * Test job for each PR: "test_develop_pr_core":https://ci.theforeman.org/job/test_develop_pr_core/

Reviewer 2: ```suggestion
* Test job for each PR: [test_develop_pr_core](https://ci.theforeman.org/job/test_develop_pr_core/)
```

PR jobs should be set up identically to primary branch tests, except for the SCM (which checks out `${sha1}`) and to add the GHPRB trigger (see the `github_pr` macro in JJB).

To set up the hook, an org/repo admin goes to the repository settings, then Webhooks & Services and adds a webhook with these settings:

* Payload URL: `https://ci.theforeman.org/ghprbhook/`
* Content type: `application/json`
* Secret: _redacted_
* Events: _Let me select individual events_, _Pull request_, _Issue comment_

An org admin must then change the org teams:

* Add the repository to the "Bots team":https://github.com/orgs/theforeman/teams/bots/repositories with +write+ access

### CentOS QA service

CentOS offer access to their "QA CI infrastructure":https://wiki.centos.org/QaWiki/CI to Foreman and other projects, which gives on-demand access to run tests on physical hardware running CentOS.  It consists of three main components:

* "ci.centos.org":https://ci.centos.org - a Jenkins instance which we can manage jobs on

> * foreman`slave01 - a user account on a slave which our jobs all run on

Reviewer 2: ```suggestion
* foreman@slave01 - a user account on a slave which our jobs all run on
```

# Jenkins

The Foreman project maintains its own Jenkins instance for continuous integration at https://ci.theforeman.org/

It runs the following types of tests:

* Unit tests on the main develop and all supported stable branches of core Foreman projects
* Pull request tests against core Foreman projects
* Unit and pull request tests for many Foreman plugins
* Nightly and release package building and release processes
* Package smoke and integration testing
* Package jobs for Pulpcore

## Quick reference for maintainers

Current PR test jobs (used on Foreman itself) support these commands:

* `ok to test` - run tests for an unknown user, if the code within the patch is not malicious
* `add to PR whitelist` - automatically run all PRs in future from this user
* `[test STATUS-NAME]`, e.g. `[test foreman]` to re-run a particular set of tests

Older PR tests, those that use the status name "default", supports this command:

> * `[test]` - run tests for an unknown user, or re-run all tests

Reviewer: I think we got rid of those and now only use GHPRB. The old PR Processor code has been removed so I can't imagine this is still relevant.

## Quick reference for plugin maintainers

### Foreman plugin testing

Foreman plugins are tested by adding the plugin to a Foreman checkout and running core tests, so it checks that existing behaviours still work and new plugin tests are run too.  The "test_plugin_matrix job":https://ci.theforeman.org/job/test_plugin_matrix/ copies the core jobs, but adds a plugin from a given git repo/branch and is usually used to test plugins in a generic way.

Each plugin should have a job defined in JJB that calls test_plugin_matrix here: https://ci.theforeman.org/view/Plugins/

#### Foreman plugin PR testing

To test pull requests, a separate job is used that also takes the PR details: https://ci.theforeman.org/view/Plugins/job/test_plugin_pull_request/

> To set this up for a plugin, make the template + manifest addition as noted in the PR section below to configure the PR scanner.  Also enable the prprocessor webhook on the GitHub repo for immediate builds.

Reviewer: There is no more PR scanner.

#### Adding a new Foreman plugin

For a plugin "foreman_example", first create a job that tests the main (master or develop) branch.

* ensure plugin tests (if any) run when `rake jenkins:unit` is called, see "the example plugin":https://github.com/theforeman/foreman_plugin_template/ and "testing a plugin":https://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Testing for help
* create a foreman_example.yaml file in foreman-infra/JJB
  * https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files/theforeman.org/yaml/jobs/plugins
  * This will create a "test_plugin_foreman_example_master" job in Jenkins to test the master branch.
* ensure the job is green by fixing bugs, installing dependencies etc.
* add hook to GitHub repo, see "GitHub repo hook":https://projects.theforeman.org/projects/foreman/wiki/Jenkins#Testing-develop

Next, set up PR testing for the plugin:

> * add template to foreman-infra/puppet/modules/slave/templates, update branch, job names etc.

Reviewer: I'm not sure what this means, but pretty sure this can be dropped as well.

> * add project to slave::pr_test_config list in foreman-infra/puppet/modules/slave/manifests/init.pp
> * add project to [[PrProcessor]] "repos.yaml":https://github.com/theforeman/prprocessor/blob/master/config/repos.yaml config file with "pr_scanner: true"

Reviewer: No longer needed
```suggestion
```

An org admin must then:

> * add [[PrProcessor]] webhook to GitHub repo

Reviewer: This is no longer needed

# Jenkins

> The Foreman project maintains its own Jenkins instance for continuous integration at http://ci.theforeman.org/

Reviewer 2: ```suggestion
The Foreman project maintains its own Jenkins instance for continuous integration at https://ci.theforeman.org/
```

It runs the following types of tests:

> * Unit tests on the main develop and recent stable branches of core Foreman projects

Reviewer 2: ```suggestion
* Unit tests on the main develop and all supported stable branches of core Foreman projects
```

* Pull request tests against core Foreman projects
* Unit and pull request tests for many Foreman plugins
* Nightly and release package building and release processes

> * Package smoke and integration testing ("systest")

Reviewer 2: I'd drop "systest" as we don't use that word anywhere anymore.

Should list also mention Katello and Pulpcore individually?

## Quick reference for maintainers

Current PR test jobs (used on Foreman itself) support these commands:

> * @ok to test@ - run tests for an unknown user, if the code within the patch is not malicious
> * @add to PR whitelist@ - automatically run all PRs in future from this user
> * @[test STATUS-NAME]@, e.g. @[test foreman]@ to re-run a particular set of tests

Reviewer 2: ```suggestion
* `ok to test` - run tests for an unknown user, if the code within the patch is not malicious
* `add to PR whitelist` - automatically run all PRs in future from this user
* `[test STATUS-NAME]`, e.g. `[test foreman]` to re-run a particular set of tests
```

Older PR tests, those that use the status name "default", supports this command:

> * @[test]@ - run tests for an unknown user, or re-run all tests

Reviewer 2: ```suggestion
* `[test]` - run tests for an unknown user, or re-run all tests
```

is that still true?

## Quick reference for plugin maintainers

### Foreman plugin testing

> Foreman plugins are tested by adding the plugin to a Foreman checkout and running core tests, so it checks that existing behaviours still work and new plugin tests are run too.  The "test_plugin_matrix job":http://ci.theforeman.org/job/test_plugin_matrix/ copies the core jobs, but adds a plugin from a given git repo/branch and is usually used to test plugins in a generic way.

Reviewer 2: ```suggestion
Foreman plugins are tested by adding the plugin to a Foreman checkout and running core tests, so it checks that existing behaviours still work and new plugin tests are run too.  The [test_plugin_matrix job](https://ci.theforeman.org/job/test_plugin_matrix/) copies the core jobs, but adds a plugin from a given git repo/branch and is usually used to test plugins in a generic way.
```

# View the repository settings
# Click _Webhooks & Services_
# _Add service_ dropdown, select _Jenkins (Git plugin)_
# For the URL, enter @http://ci.theforeman.org/@

### Pull request testing

Core Foreman projects have GitHub pull requests tested on our Jenkins instance so it's identical to the way we test the primary development branches themselves.  Simpler and quieter projects (such as installer modules) should use Travis CI which supports PR testing and reduces the load on our own infrastructure.

Every project that needs PR testing has at least two Jenkins jobs.  Taking core Foreman as an example:

* Test job for the main development branch (develop or master): "test_develop":http://ci.theforeman.org/job/test_develop/
* Test job for each PR: "test_develop_pr_core":http://ci.theforeman.org/job/test_develop_pr_core/

The results from these PR jobs are only stored for a few weeks, sufficient for reviews. The PR jobs are set up in a few different ways, as it's changed over time:

# _Version 3_ jobs use the _GitHub Pull Request Builder_ (GHPRB) and merge the branch to the project branch
# _Version 2_ jobs use the _PR scanner_ and merge the branch to the primary project branch

> # _Version 1_ jobs use the _PR scanner_ and use @git apply@ with the patch from the PR

Reviewer 2: do v2 and v1 still exist?

# Branches to build: @develop@
# Additional Behaviours > Prune stale remote-tracking branches
# Additional Behaviours > Wipe out repository & force clone

As the first build step, add:

<pre>
#!/bin/bash -ex
if [ -n "${pr_git_url}" ]; then
  git remote add pr ${pr_git_url}
  git fetch pr
  git merge pr/${pr_git_ref}
fi
</pre>

#### Adding a new generic project

For a project "foo":

# create a test_foo job that tests the primary development branch, enable IRC notifications in post build
# clone to test_foo_pull_request
#* without IRC notifications
#* add pr_number parameter
#* add initial build step to download the patch and apply
# add template to foreman-infra/puppet/modules/slave/templates, update branch, job names etc.
# add project to slave::pr_test_config list in foreman-infra/puppet/modules/slave/manifests/init.pp
# add project to [[PrProcessor]] "repos.yaml":https://github.com/theforeman/prprocessor/blob/master/config/repos.yaml config file with "pr_scanner: true"

An org admin must then:

# add [[PrProcessor]] webhook to GitHub repo
# add the repo to the "Bots team":https://github.com/orgs/theforeman/teams/bots/repositories with +write+ access

### CentOS QA service

> * _This change is in development, not currently in use._

Reviewer 2: not true!

# "Duffy":http://wiki.centos.org/QaWiki/CI/Duffy - on-demand provisioning of physical test servers running CentOS, available from foreman@slave01 jobs

## Managing jobs

Jenkins itself is deployed onto one master VM from foreman-infra.  Jobs are maintained in two ways:

# Updated by hand - some developers have accounts to log in and modify jobs
# Jenkins Job Builder - a set of YAML configuration files in foreman-infra that generate jobs

New jobs should be written via Jenkins Job Builder if possible.

### Jenkins Job Builder

"Jenkins Job Builder":http://docs.openstack.org/infra/jenkins-job-builder/ (JJB) is an OpenStack tool for generating Jenkins job definitions (an XML file) from a set of YAML job descriptions, which we store in version control.

* "JJB YAML files in foreman-infra":https://github.com/theforeman/foreman-infra/tree/master/puppet/modules/jenkins_job_builder/files

Puppet deploys these onto our Jenkins server (a recursive file copy) and when they change, it runs the JJB tool to update the jobs in the live instance.  It also refreshes them daily to overwrite manual changes.

We use a fork of JJB with some extra support for plugins we use:

> * "theforeman/jenkins-job-builder, local branch":https://github.com/theforeman/jenkins-job-builder/tree/local

Reviewer 2: we don't anymore
